### PR TITLE
perf(style): skip descendant cascade when no child style is keyed on the changed state bits

### PR DESCRIPTION
**Pull request by zayaanquilt**
### Summary

Reworked based on @kisvegabor's and the @copilot-pull-request-reviewer feedback on the original approach (correctly flagged as a semantic break — `lv_table` pressed-state suppression and list `LV_PART_MAIN` scrolled styling both consume `LV_STATE_SCROLLED` regardless of scrollbar mode). This version attacks the real root cause: the descendant style cascade itself, not the state toggle.

### Root cause

`update_obj_state` already short-circuits via `lv_obj_style_state_compare` when the parent's own styles don't change. But once the cascade fires, `refresh_children_style` unconditionally calls `lv_obj_invalidate` + `lv_obj_send_event(..., LV_EVENT_STYLE_CHANGED, ...)` on every descendant — regardless of whether the descendant has any style whose selector actually uses the state bit(s) that flipped. The per-widget `LV_EVENT_STYLE_CHANGED` handler work dominates on deep trees; most descendants do no observable work because nothing they render depends on the changed bits.

### Approach

1. New entry point `lv_obj_style_refresh_on_state_change(obj, changed_bits)` sets a file-static `refresh_cascade_bits = changed_bits`, runs the existing `lv_obj_refresh_style(obj, LV_PART_ANY, LV_STYLE_PROP_ANY)`, then clears. Asserts `refresh_cascade_bits == 0` on entry to catch accidental nesting — LVGL is single-threaded at this boundary.

2. `refresh_children_style` consults the filter: when non-zero, it invalidates + dispatches only if the child has at least one non-transition style whose selector state intersects the filter bits (new static `obj_has_selector_on_bits`). Recursion into grandchildren stays unconditional — a grandchild may have state-keyed styles its immediate parent lacks.

3. `update_obj_state` routes its two state-transition calls (`CMP_DIFF_REDRAW` and `CMP_DIFF_LAYOUT`) through the new entry point with `prev_state ^ new_state` as the changed bits. `CMP_DIFF_DRAW_PAD` is untouched (no cascade there).

4. Defensive inheritable-prop fallback: before activating the filter, the entry point walks `obj->styles[]`. If any non-transition style whose selector intersects `changed_bits` carries `LV_STYLE_PROP_FLAG_INHERITABLE` (e.g. `LV_STYLE_TEXT_COLOR`), the filter is disabled for that refresh — descendants inheriting the prop need the repaint even without their own state-keyed selector.

### Why this preserves semantics

- `LV_STATE_SCROLLED` (and every other state bit) toggles unconditionally, as in master. `lv_table`'s pressed-state suppression and any list `LV_PART_MAIN | LV_STATE_SCROLLED` styling see the state exactly as before.
- The parent object itself always goes through `lv_obj_refresh_style` — nothing changes about its own invalidate / layer-type / transition handling.
- Descendants with state-keyed styles still receive `LV_EVENT_STYLE_CHANGED`.
- Descendants without state-keyed styles skip the dispatch — they had no style-dependent response to give, so the event was no-op work.
- The inheritable-prop fallback covers the "parent has state-keyed inheritable style" case conservatively.

### Generalizes

This applies to every state bit, not just `LV_STATE_SCROLLED`. Pressed / Focused / Edited / Hovered transitions on deep trees benefit identically.

### Measurement

The cascade cost was previously established on this PR: on an embedded Linux device (454×454 single-buffer fbdev, `-O2`) with a tileview containing ~100+ widgets per tile, `LV_EVENT_STYLE_CHANGED` was 190 ms / 4638 calls (~31% of scroll self-time) over a 613 ms scroll window. That work was the descendant cascade walking a subtree with no state-keyed consumers — exactly what this revision's filter skips.

Visually re-verified on the same hardware with the new patch: scroll UX matches the previously-measured behaviour. Happy to re-run the profiler with an instrumented cascade-hit counter if reviewers want empirical confirmation that the inheritable-prop fallback isn't triggering on the hot path. LVGL's CI benchmark bot will also cover this on the force-push automatically.

### CODE REVIEW

**Comment by zayaanquilt** [comment]
Good catch — fixed in 22ae5a4. Made the handling asymmetric: the `SCROLL_BEGIN` guard stays (that's where the cascade cost lives), but `SCROLL_END` now removes unconditionally. When the state isn't set, `lv_obj_remove_state` early-returns at the bitmask check without cascading, so the perf win is preserved while handling the mid-scroll mode-change edge case.

File: src/core/lv_obj.c [hunk 1]
```diff
@@ -1053,7 +1053,13 @@ static void lv_obj_event(const lv_obj_class_t * class_p, lv_event_t * e)
...
     }
     else if(code == LV_EVENT_SCROLL_BEGIN) {
-        lv_obj_add_state(obj, LV_STATE_SCROLLED);
+        /*Only toggle LV_STATE_SCROLLED when a scrollbar is actually being
+         *shown. Otherwise the state change triggers lv_obj_style_state_compare
+         *-> refresh_children_style which recursively fires LV_EVENT_STYLE_CHANGED
+         *on every descendant for a style nobody renders.*/
+        if(lv_obj_get_scrollbar_mode(obj) != LV_SCROLLBAR_MODE_OFF) {
+            lv_obj_add_state(obj, LV_STATE_SCROLLED);
+        }
```
[the remark this answers is not in the document]
Comment (kisvegabor) [reply src/core/lv_obj.c]: Actually this is true. An other example is when the a list itself `LV_PART_MAIN` has a different style when it's scrolled.

**Comment by zayaanquilt** [comment]
Force-pushed a reworked approach based on the review feedback — the cascade is now filtered in `refresh_children_style` instead of gating `LV_STATE_SCROLLED`, so the state toggle stays unconditional and `lv_table` / list `LV_PART_MAIN` semantics are preserved. Updated PR description above has the full rationale. Longer reply to each comment coming.

File: src/core/lv_obj_style.c [hunk 2]
```diff
@@ -933,12 +941,55 @@ static void refresh_children_style(lv_obj_t * obj)
...
+        if((sel_state & changed_bits) == 0) continue;
+        if(style_has_flag(obj->styles[i].style, LV_STYLE_PROP_FLAG_INHERITABLE)) {
+            bits = 0;
+            break;
+        }
     }
+
+    refresh_cascade_bits = bits;
+    lv_obj_refresh_style(obj, LV_PART_ANY, LV_STYLE_PROP_ANY);
+    refresh_cascade_bits = 0;
```
[the remark this answers is not in the document]
Comment (zayaanquilt) [reply src/core/lv_obj_style.c]: Done in 1ec5cee — save the previous `refresh_cascade_bits` and restore it after the inner refresh as you suggested. Inner cascade uses its own filter; outer cascade resumes with its previous filter intact.

File: src/core/lv_obj_style.c [hunk 3]
```diff
@@ -933,12 +941,55 @@ static void refresh_children_style(lv_obj_t * obj)
...
+        if(sel_state & state_bits) return true;
+    }
+    return false;
+}
+
+void lv_obj_style_refresh_on_state_change(lv_obj_t * obj, lv_state_t changed_bits)
+{
+    /*Single-threaded LVGL invariant: no reentry while a state refresh is in
+     *progress. Catches accidental nesting.*/
+    LV_ASSERT(refresh_cascade_bits == 0);
```
[the remark this answers is not in the document]
Comment (zayaanquilt) [reply src/core/lv_obj_style.c]: Right — reentry is legal (event handlers in widget code can mutate state during the cascade, and CI test_textarea/test_demo_stress hit exactly this). Dropped the assert in 1ec5cee and made the entry point reentrant via save/restore of the previous `refresh_cascade_bits`.

File: src/core/lv_obj_style_private.h [hunk 4]
```diff
@@ -86,6 +86,18 @@ lv_style_state_cmp_t lv_obj_style_state_compare(lv_obj_t * obj, lv_state_t state
...
+ * Same as `lv_obj_refresh_style(obj, LV_PART_ANY, LV_STYLE_PROP_ANY)` but the
+ * descendant cascade (`refresh_children_style`) skips dispatching
+ * `LV_EVENT_STYLE_CHANGED` to subtrees with no styles keyed on the changed
+ * state bits. Intended for the state-transition path in `update_obj_state`
+ * where the only thing that changed is a set of state bits and descendants
+ * without selectors on those bits can't render differently.
+ * @param obj           the object whose state changed
+ * @param changed_bits  `prev_state ^ new_state` — the bits that flipped
+ */
+void lv_obj_style_refresh_on_state_change(lv_obj_t * obj, lv_state_t changed_bits);
```
Comment (kisvegabor) [inline src/core/lv_obj_style_private.h]: We should handle this internally, hidden from the user. Please make it a `static` function.

File: src/core/lv_obj_style.c [hunk 5]
```diff
@@ -933,12 +941,56 @@ static void refresh_children_style(lv_obj_t * obj)
...
+{
+    for(uint32_t i = 0; i < obj->style_cnt; i++) {
+        if(obj->styles[i].is_trans) continue;
+        lv_state_t sel_state = lv_obj_style_get_selector_state(obj->styles[i].selector);
+        if(sel_state & state_bits) return true;
     }
+    return false;
+}
+
+void lv_obj_style_refresh_on_state_change(lv_obj_t * obj, lv_state_t changed_bits)
```
Comment (kisvegabor) [inline src/core/lv_obj_style.c]:
```suggestion
void lv_obj_style_refresh_on_state_change(lv_obj_t * obj, lv_state_t state_changed)
```

File: src/core/lv_obj_style.c [hunk 6]
```diff
@@ -933,12 +941,56 @@ static void refresh_children_style(lv_obj_t * obj)
...
+        }
+    }
+
+    /*Save and restore: an event handler dispatched during the cascade may
+     *legally trigger another state change, re-entering this function. We
+     *want the inner cascade to use its own filter and the outer cascade to
+     *resume with its previous filter intact, not be cleared to "no filter."*/
+    lv_state_t prev_refresh_cascade_bits = refresh_cascade_bits;
+    refresh_cascade_bits = bits;
+    lv_obj_refresh_style(obj, LV_PART_ANY, LV_STYLE_PROP_ANY);
```
Comment (kisvegabor) [inline src/core/lv_obj_style.c]: This function is used only internally, and it's really not meant to be used by the users. I suggest making it  private (moving it's declaration to `lv_obj_style_private.h`)  and add an extra `state changed` parameter to avoid using a global variable.

File: src/core/lv_obj_style.c [hunk 7]
```diff
@@ -933,12 +941,56 @@ static void refresh_children_style(lv_obj_t * obj)
...
+        if(obj->styles[i].is_trans) continue;
+        lv_state_t sel_state = lv_obj_style_get_selector_state(obj->styles[i].selector);
+        if(sel_state & state_bits) return true;
     }
+    return false;
+}
+
+void lv_obj_style_refresh_on_state_change(lv_obj_t * obj, lv_state_t changed_bits)
+{
+    lv_state_t bits = changed_bits;
```
Comment (kisvegabor) [inline src/core/lv_obj_style.c]: Use more descriptive name here.

**Review by kisvegabor — COMMENTED** [review]
Thank you for the update. It's a really complex part  and I couldn't fully follow the flow. Please address my comments to make the code easier to understand, and also add a few tests to check the edge cases.
